cpp/common/src/codingstandards/cpp/deadcode/UnusedObjects.qll

```diff
@@ -0,0 +1,176 @@
+import cpp
+import codingstandards.cpp.alertreporting.HoldsForAllCopies
+import codingstandards.cpp.alertreporting.DeduplicateMacroResults
+
+/**
+ * An unused object definition is an object, meaning a place in memory, whose definition could be
+ * removed and the program would still compile.
+ *
+ * Technically, parameters may be considered objects, but they are covered by their own rule.
+ * Similarly, members of structs are an addressable place in memory, and may be considered objects.
+ * However, the member declaration is nothing but a layout offset, which is not an object.
+ *
+ * This therefore only reports variables (local or top level) which have a definition, and are
+ * unused.
+ */
+class UnusedObjectDefinition extends VariableDeclarationEntry {
+  UnusedObjectDefinition() {
+    not exists(VariableAccess access | access.getTarget() = getVariable()) and
+    getVariable().getDefinition() = this and
+    not this instanceof ParameterDeclarationEntry and
+    not getVariable() instanceof MemberVariable
+  }
```

Comment:
I think we should review, and potentially integrate, the queries from previous standards that look for unused local variables.

 * [UnusedLocalVariable.ql](https://github.com/github/codeql-coding-standards/blob/main/cpp/autosar/src/rules/M0-1-3/UnusedLocalVariable.ql)
 * [UnusedVariables.qll](https://github.com/github/codeql-coding-standards/blob/main/cpp/common/src/codingstandards/cpp/deadcode/UnusedVariables.qll)

These are both C++ queries, so not everything will be applicable, but notably:
 * Ignoring results in functions that use ASM statements (as we cannot determine if the assembly instructions access the the variable).
 * Consider ignoring results in functions with error expressions.
 * Excluding compiler generated variables.

Reply:
Integrated, thanks!